bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerHandle.java

```diff
@@ -1100,6 +1100,68 @@ public void asyncAddEntry(final long entryId, final byte[] data, final int offse
         cb.addCompleteWithLatency(BKException.Code.IllegalOpException, LedgerHandle.this, entryId, 0, ctx);
     }
 
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public CompletableFuture<Void> force() {
+        CompletableFuture<Void> result = new CompletableFuture<>();
+        ForceLedgerOp op = new ForceLedgerOp(this, result);
+        boolean wasClosed = false;
+        synchronized (this) {
+            // synchronized on this to ensure that
+            // the ledger isn't closed between checking and
+            // updating lastAddPushed
+            if (metadata.isClosed()) {
+                wasClosed = true;
+            }
+        }
+
+        if (wasClosed) {
+            // make sure the callback is triggered in main worker pool
+            try {
+                bk.getMainWorkerPool().submit(new SafeRunnable() {
+                    @Override
+                    public void safeRun() {
+                        LOG.warn("Force() attempted on a closed ledger: {}", ledgerId);
+                        result.completeExceptionally(new BKException.BKLedgerClosedException());
```

Comment:
Should we do an implicit force() on close()? We are failing the force() if the ledger is closed, but the close() API doesn't appear to guarantee the force(). So what happens to applications that failed to force() before close()?

Reply:
We have discussed about this point. Final answer was to leave this up to the application.
Filesystems usually do to fsync on close.

I have a test case for this scenario. If you do not perform force() LAC does not advance and metadata are sealed with the current LAC, which could be -1 if no force() has been issued and completed with success. This is consistent with what readers had seen, that is a LAC which never has been updated.
I see this is not obvious and should be clearly stated in high level docs about BP14, not only in javadocs

Reply:
We have discussed about this point. Final answer was to leave this up to the application.
Filesystems usually do to fsync on close.

I have a test case for this scenario. If you do not perform force() LAC does not advance and metadata are sealed with the current LAC, which could be -1 if no force() has been issued and completed with success. This is consistent with what readers had seen, that is a LAC which never has been updated.
I see this is not obvious and should be clearly stated in high level docs about BP14, not only in javadocs